exercises/dnd-character/test/Tests.hs

```diff
@@ -0,0 +1,72 @@
+{-# OPTIONS_GHC -fno-warn-type-defaults #-}
+{-# LANGUAGE RecordWildCards #-}
+
+import Data.Foldable     (for_)
+import Test.Hspec        (Spec, describe, it, shouldBe)
+import Test.Hspec.Runner (configFastFail, defaultConfig, hspecWith)
+import Test.QuickCheck
+
+import DND (Character(..), ability, modifier, character)
+
+data Case = Case { input    :: Int
+                 , expected :: Int
+                 }
+
+type Ability = Character -> Int
+
+main :: IO ()
+main = hspecWith defaultConfig {configFastFail = True} specs
+
+specs :: Spec
+specs = do
+  describe "modifier" $ for_ cases test
+
+  describe "ability" $ it "generates values within range" $
+    property (forAll ability abilityWithinRange)
+
+  describe "character" $ it "generates valid characters" $
+    property (forAll character characterIsValid)
+  where
+    test Case{..} = it description assertion
+      where
+        description = "computes the modifier for " ++ showInput input ++ " to be " ++ showMod expected
+        assertion = modifier input `shouldBe` expected
+        showInput n
```

Comment:
optional: if you so desire, it would be possible to achieve these six lines with a printf

Reply:
Done.